src/oic/extension/provider.py

```diff
@@ -640,19 +640,9 @@ def client_credentials_grant_type(self, areq):
         return Response(atr.to_json(), content="application/json")
 
     def password_grant_type(self, areq):
-        _at = self.token_handler.get_access_token(areq['client_id'],
-                                                  scope=areq['scope'],
-                                                  grant_type='password')
-        _info = self.token_handler.token_factory.get_info(_at)
-        try:
-            _rt = self.token_handler.get_refresh_token(
-                self.baseurl, _info['access_token'], 'password')
-        except NotAllowed:
-            atr = self.do_access_token_response(_at, _info, areq['state'])
-        else:
-            atr = self.do_access_token_response(_at, _info, areq['state'], _rt)
-
-        return Response(atr.to_json(), content="application/json")
+        _at = self.sdb.upgrade_to_token(key=areq['password'], issue_refresh=True)
```

Comment:
This looks like it would issue a refresh token to a public client that may not have 'offline_access' scope.

Probably needs some scope validation too (like it is done in oic.provider for the code_grant_type to determine if a refresh token should be issued.)

Reply:
I am not 100% sure on this. The specs are a little bit vague in this respect and there is no mention of the scoping https://tools.ietf.org/html/rfc6749#section-4.3.3.

The issuing of the access credentials is entirely out of scope of the specs, so I am not sure where the `offline_access` scope would come in. We can add a flag to the ClientDB just for this purpose.

Reply:
See https://tools.ietf.org/html/rfc6749#section-4.3.2 where the request is described. The rules for offline_access scope would come from https://openid.net/specs/openid-connect-core-1_0.html#OfflineAccess.

But you are right, the spec is vague and the OpenID connect thing mentions a MAY for issuing the refresh token anyway.

I would pick a default policy of deny the refresh token and maybe add a configurable option to grant it anyway, if configured that way.

Reply:
Yes, the extension is OAuth2 only. My understanding of the specs is that we should either issue or not issue the refresh but not distinguish between clients.

Reply:
Ok, that invalidates my comment about offline_access scope.

The spec is vague, but hints at some distinction between clients in section 2.1.

But ok, issuing a refresh_token seems fine in this case.

Reply:
OK, we will keep it this way then.